File: src/types/project.ts
/**
 * Project Types - 项目和会话类型定义
 * Story 2.8: Dashboard 项目列表
 *
 * 定义项目聚合和会话相关的数据结构
 */

/**
 * 会话来源类型
 */
export type SessionSource = "claude" | "gemini" | "cursor" | "unknown";

/**
 * 会话摘要信息 (Rust: SessionSummary)
 * 用于项目列表中的会话展示
 */
export interface Session {
  /** 会话唯一标识 */
  id: string;
  /** 会话来源 */
  source: SessionSource;
  /** 创建时间 (ISO 8601 字符串) */
  created_at: string;
  /** 更新时间 (ISO 8601 字符串) */
  updated_at: string;
  /** 消息数量 */
  message_count: number;
  /** 会话是否为空 (Story 2.29) */
  is_empty?: boolean;
  /** 会话标题 (Story 1.12) */
  title?: string;
  /** 原始工作目录 (Story 1.12: 导入时记录，不可修改) */
  original_cwd?: string;
  /** 来源上下文 (Story 1.12: 导入源特定的元数据) */
  source_context?: SourceContext;
}

/**
 * 路径类型 (Story 1.12)
 *
 * 用于分类路径的类型，决定如何验证和显示路径。
 */
export type PathType = "local" | "virtual" | "remote";

/**
 * 项目信息 (匹配 Rust: Project，snake_case)
 * 从 list_projects 返回的数据
 */
export interface Project {
  /** 项目唯一标识 */
  id: string;
  /** 项目名称 (目录名) */
  name: string;
  /** 项目路径 (工作目录) */
  cwd: string;
  /** 会话数量 */
  session_count: number;
  /** 非空会话数量 - Story 2.29 V2 */
  non_empty_session_count?: number;
  /** 创建时间 (ISO 8601 字符串) */
  created_at: string;
  /** 最后活动时间 (ISO 8601 字符串) */
  last_activity: string;
  /** Git 仓库根路径 (如果检测到) */
  git_repo_path: string | null;
  /** 是否关联 Git 仓库 */
  has_git_repo: boolean;
  /** Git 远程 URL (Story 1.9: 用于跨路径项目聚合) */
  git_remote_url: string | null;
  /** 项目是否为空 (所有会话都是空会话) - Story 2.29 V2 */
  is_empty?: boolean;
  /** 路径类型 (Story 1.12: local/virtual/remote) */
  path_type?: PathType;
  /** 本机路径是否存在 (Story 1.12: 仅对 local 类型有效) */
  path_exists?: boolean;
}

/**
 * 代表性文件信息 (匹配 Rust: RepresentativeFile)
 * Story 2.11: 用于显示项目初始代码状态
 */
export interface RepresentativeFile {
  /** 文件路径 (相对于仓库根目录) */
  path: string;
  /** 文件内容 */
  content: string;
  /** 检测到的编程语言 */
  language: string;
}

/**
 * 快照结果 (匹配 Rust: SnapshotResult)
 * 用于获取 Git HEAD 版本的文件内容
 */
export interface SnapshotResult {
  /** 文件内容 */
  content: string;
  /** Commit Hash */
  commit_hash: string;
  /** Commit 消息 */
  commit_message: string;
  /** Commit 时间戳 (Unix seconds) */
  commit_timestamp: number;
}

/**
 * 导入结果 (匹配 Rust: ImportResult)
 */
export interface ImportResult {
  /** 成功导入的会话数 */
  imported_count: number;
  /** 跳过的重复会话数 */
  skipped_count: number;
  /** 新创建的项目数 */
  new_projects_count: number;
  /** 错误列表 */
  errors: string[];
}

// ============================================================================
// Story 1.12: View-based Project Aggregation Types
// ============================================================================

/**
 * 项目路径映射 (匹配 Rust: ProjectPath)
 *
 * 一个项目可以关联多个路径，实现灵活的项目聚合。
 */
export interface ProjectPath {
  /** 唯一标识 (UUID) */
  id: string;
  /** 所属项目 ID */
  project_id: string;
  /** 路径 (已标准化) */
  path: string;
  /** 是否为主路径 */
  is_primary: boolean;
  /** 创建时间 (ISO 8601 字符串) */
  created_at: string;
}

/**
 * 会话手动绑定 (匹配 Rust: SessionBinding)
 *
 * 允许用户手动将会话绑定到特定项目，
 * 优先级高于基于路径的自动匹配。
 */
export interface SessionBinding {
  /** 会话 ID */
  session_id: string;
  /** 绑定的项目 ID */
  project_id: string;
  /** 绑定时间 (ISO 8601 字符串) */
  bound_at: string;
}

/**
 * 来源上下文 (匹配 Rust: SourceContext)
 *
 * 存储导入源特定的元数据，用于标识会话来源。
 * 导入后不可变，用于调试/审计。
 */
export interface SourceContext {
  /** 原始文件路径 */
  file_path?: string;
  /** Claude Code: 编码的项目路径 */
  project_path_encoded?: string;
  /** Gemini CLI: 项目哈希 */
  project_hash?: string;
  /** Gemini CLI: 会话文件名 */
  session_filename?: string;
  /** Cursor: 工作区 ID */
  workspace_id?: string;
  /** Cursor: 工作区存储路径 */
  workspace_path?: string;
}

/**
 * 逻辑项目统计 (匹配 Rust: LogicalProjectStats)
 *
 * 按物理路径聚合的项目统计信息。
 * 用于视图层显示"逻辑项目"，将来自不同导入源的会话聚合在一起。
 *
 * Story 1.12 Phase 5: 增强数据结构
 */
export interface LogicalProjectStats {
  /** 物理路径 (已标准化) */
  physical_path: string;
  /** 关联此路径的项目数量 */
  project_count: number;
  /** 关联此路径的所有项目 ID */
  project_ids: string[];
  /** 所有项目的会话总数 */
  total_sessions: number;
  /** 最近活动时间 (ISO 8601 字符串) */
  last_activity: string;
  /** 显示名称 (从路径提取) - Task 8.1 */
  display_name: string;
  /** 路径类型: local/virtual/remote - Task 8.2 */
  path_type: PathType;
  /** 本机路径是否存在 (仅 local 类型有效) - Task 8.3 */
  path_exists: boolean;
  /** 是否需要关联真实路径 - Task 8.4
   * True if path_type is "virtual" or (path_type is "local" AND path_exists is false)
   */
  needs_association: boolean;
  /** 是否关联 Git 仓库 (聚合自关联项目) - Task 17 */
  has_git_repo: boolean;
}

File: src-tauri/src/models/project.rs
//! Project and session summary data models
//!
//! Defines the Project and SessionSummary structures for representing
//! aggregated project data and lightweight session listings.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::session::SessionSource;

/// Path type classification (Story 1.12)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum PathType {
    /// Local filesystem path (e.g., /home/user/project)
    #[default]
    Local,
    /// Virtual placeholder path (e.g., gemini-project:xxx, placeholder:xxx)
    Virtual,
    /// Remote path (e.g., github.com/user/repo)
    Remote,
}

impl PathType {
    /// Convert from string representation
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "virtual" => PathType::Virtual,
            "remote" => PathType::Remote,
            _ => PathType::Local,
        }
    }

    /// Convert to string representation
    pub fn as_str(&self) -> &'static str {
        match self {
            PathType::Local => "local",
            PathType::Virtual => "virtual",
            PathType::Remote => "remote",
        }
    }
}

/// Project data representing an aggregated working directory
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct Project {
    /// Project unique ID (UUID)
    pub id: String,
    /// Project name (directory name extracted from cwd)
    pub name: String,
    /// Working directory absolute path (primary path, kept for backward compatibility)
    pub cwd: String,
    /// Number of sessions in this project
    pub session_count: u32,
    /// Number of non-empty sessions in this project (Story 2.29 V2)
    /// Used for displaying filtered count when "hide empty sessions" is enabled
    #[serde(default)]
    pub non_empty_session_count: u32,
    /// First import time
    pub created_at: DateTime<Utc>,
    /// Last activity time (latest session's updated_at)
    pub last_activity: DateTime<Utc>,
    /// Git repository root path (if detected)
    pub git_repo_path: Option<String>,
    /// Whether this project has an associated Git repository
    pub has_git_repo: bool,
    /// Git remote URL for project identification (Story 1.9)
    /// Used for cross-path project aggregation (same repo, different paths)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub git_remote_url: Option<String>,
    /// Whether this project is empty (all sessions are empty)
    /// Story 2.29 V2: Used for filtering empty projects in the UI
    #[serde(default)]
    pub is_empty: bool,
    /// Path type classification (Story 1.12)
    /// Determines how the path should be validated and displayed
    #[serde(default)]
    pub path_type: PathType,
    /// Whether the local path exists on the filesystem (Story 1.12)
    /// Only meaningful for PathType::Local
    #[serde(default = "default_path_exists")]
    pub path_exists: bool,
}

fn default_path_exists() -> bool {
    true
}

/// Project path mapping (Story 1.12)
///
/// A project can have multiple paths associated with it.
/// This enables flexible project aggregation from different sources.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProjectPath {
    /// Unique ID (UUID)
    pub id: String,
    /// Project ID this path belongs to
    pub project_id: String,
    /// The path (normalized)
    pub path: String,
    /// Whether this is the primary path for the project
    pub is_primary: bool,
    /// When this path was associated with the project
    pub created_at: DateTime<Utc>,
}

/// Session-to-project manual binding (Story 1.12)
///
/// Allows users to manually bind a session to a specific project,
/// overriding the automatic path-based matching.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct SessionBinding {
    /// Session ID
    pub session_id: String,
    /// Project ID the session is bound to
    pub project_id: String,
    /// When the binding was created
    pub bound_at: DateTime<Utc>,
}

/// Source context for session import (Story 1.12)
///
/// Stores source-specific metadata that helps identify the session origin.
/// This is immutable after import and used for debugging/auditing.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct SourceContext {
    /// Original file path from the source
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_path: Option<String>,

    /// Claude Code: Encoded project path from directory name
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_path_encoded: Option<String>,

    /// Gemini CLI: Project hash from the tmp directory
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_hash: Option<String>,

    /// Gemini CLI: Session filename
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_filename: Option<String>,

    /// Cursor: Workspace ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,

    /// Cursor: Workspace storage path
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_path: Option<String>,
}

/// Lightweight session summary for listings
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct SessionSummary {
    /// Session ID
    pub id: String,
    /// Source (Claude/Gemini/Cursor)
    pub source: SessionSource,
    /// Session creation time
    pub created_at: DateTime<Utc>,
    /// Session update time
    pub updated_at: DateTime<Utc>,
    /// Message count
    pub message_count: u32,
    /// Whether this session is empty (no user AND no assistant messages)
    /// Story 2.29: Used for filtering empty projects in the UI
    pub is_empty: bool,
    /// Session title (from metadata, if available)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Original working directory at import time (Story 1.12)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub original_cwd: Option<String>,
}

/// Import result statistics
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ImportResult {
    /// Number of successfully imported sessions
    pub imported_count: u32,
    /// Number of skipped duplicate sessions
    pub skipped_count: u32,
    /// Number of newly created projects
    pub new_projects_count: u32,
    /// Import error list
    pub errors: Vec<String>,
}

impl Project {
    /// Create a new project with the given cwd
    pub fn new(id: String, cwd: String) -> Self {
        let name = extract_project_name(&cwd);
        let path_type = classify_path_type(&cwd);
        let path_exists = match path_type {
            PathType::Local => check_path_exists(&cwd),
            _ => true, // Virtual and remote paths don't need existence check
        };
        let now = Utc::now();
        Self {
            id,
            name,
            cwd,
            session_count: 0,
            non_empty_session_count: 0,
            created_at: now,
            last_activity: now,
            git_repo_path: None,
            has_git_repo: false,
            git_remote_url: None,
            is_empty: true, // New project starts as empty until sessions are added
            path_type,
            path_exists,
        }
    }

    /// Set Git repository information
    pub fn set_git_repo(&mut self, repo_path: Option<String>) {
        self.has_git_repo = repo_path.is_some();
        self.git_repo_path = repo_path;
    }

    /// Set Git remote URL (Story 1.9)
    pub fn set_git_remote_url(&mut self, url: Option<String>) {
        self.git_remote_url = url;
    }
}

/// Extract project name from cwd path
///
/// # Examples
/// ```
/// use client_lib::models::project::extract_project_name;
/// assert_eq!(extract_project_name("/Users/decker/projects/mantra"), "mantra");
/// assert_eq!(extract_project_name("/home/user/code/my-app"), "my-app");
/// ```
pub fn extract_project_name(cwd: &str) -> String {
    std::path::Path::new(cwd)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("Unknown Project")
        .to_string()
}

/// Classify path type (Story 1.12)
///
/// Determines whether a path is local, virtual, or remote.
///
/// # Examples
/// ```
/// use client_lib::models::project::{classify_path_type, PathType};
/// assert_eq!(classify_path_type("/home/user/project"), PathType::Local);
/// assert_eq!(classify_path_type("gemini-project:abc123"), PathType::Virtual);
/// assert_eq!(classify_path_type("github.com/user/repo"), PathType::Remote);
/// ```
pub fn classify_path_type(path: &str) -> PathType {
    let path = path.trim();

    // Virtual paths: placeholders and special identifiers
    if path.is_empty()
        || path == "unknown"
        || path.starts_with("gemini-project:")
        || path.starts_with("placeholder:")
    {
        return PathType::Virtual;
    }

    // Remote paths: URLs and git references
    if path.starts_with("github.com/")
        || path.starts_with("gitlab.com/")
        || path.starts_with("bitbucket.org/")
        || path.starts_with("https://")
        || path.starts_with("http://")
        || path.starts_with("git@")
    {
        return PathType::Remote;
    }

    // Default to local path
    PathType::Local
}

/// Check if a local path exists on the filesystem (Story 1.12)
///
/// # Examples
/// ```
/// use client_lib::models::project::check_path_exists;
/// // This will return true for existing paths
/// assert!(check_path_exists("/tmp"));
/// // This will return false for non-existing paths
/// assert!(!check_path_exists("/nonexistent/path/12345"));
/// ```
pub fn check_path_exists(path: &str) -> bool {
    std::path::Path::new(path).exists()
}

/// Normalize cwd path for consistent aggregation (Story 2.25)
///
/// - Removes trailing slashes
/// - Converts backslashes to forward slashes (cross-platform)
/// - Trims whitespace
///
/// # Examples
/// ```
/// use client_lib::models::project::normalize_cwd;
/// assert_eq!(normalize_cwd("/home/user/project/"), "/home/user/project");
/// assert_eq!(normalize_cwd("C:\\Users\\test\\project"), "C:/Users/test/project");
/// ```
pub fn normalize_cwd(cwd: &str) -> String {
    let normalized = cwd
        .trim()
        .replace('\\', "/"); // Cross-platform: backslashes to forward slashes

    // Remove trailing slashes (but keep root "/" or "C:/")
    let trimmed = normalized.trim_end_matches('/');

    // Handle edge cases: root paths
    if trimmed.is_empty() {
        "/".to_string()
    } else if trimmed.ends_with(':') {
        // Windows drive letter like "C:" -> "C:/"
        format!("{}/", trimmed)
    } else {
        trimmed.to_string()
    }
}

#[cfg(test)]
mod tests;
